// react-native-element-timer/lib/commonjs/CountDown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _styles = require("./styles");
var _reactNativeBackgroundTimer = _interopRequireDefault(require("react-native-background-timer"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-bitwise */

const defaultProps = {
  style: {},
  textStyle: {},
  onTimes: _seconds => {},
  onPause: _seconds => {},
  onEnd: _seconds => {}
};
const CountdownComponent = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    initialSeconds,
    style,
    textStyle,
    fontFamily,
    autoStart = false,
    formatTime = 'hh:mm:ss',
    onEnd,
    onTimes,
    onPause
  } = props;
  const interval = (0, _react.useRef)(0);
  const hours = (0, _react.useRef)(0);
  const minute = (0, _react.useRef)(0);
  const seconds = (0, _react.useRef)(0);
  const currentSeconds = (0, _react.useRef)(0);
  const [key, setKey] = (0, _react.useState)(Math.random());
  (0, _react.useImperativeHandle)(ref, () => {
    return {
      start,
      pause,
      resume,
      stop
    };
  });
  const init = (0, _react.useCallback)(() => {
    if (initialSeconds) {
      currentSeconds.current = initialSeconds;
      hours.current = ~~(currentSeconds.current / 3600);
      minute.current = ~~(currentSeconds.current % 3600 / 60);
      seconds.current = ~~currentSeconds.current % 60;
    }
    setKey(Math.random());
  }, [initialSeconds]);
  const pause = (0, _react.useCallback)(() => {
    if (onPause) {
      onPause(currentSeconds.current);
    }
    clear();
  }, []);
  const resume = () => {
    if (!interval.current) {
      timer();
    }
  };
  const stop = () => {
    if (onEnd) {
      onEnd(currentSeconds.current);
    }
    init();
    clear();
  };
  (0, _react.useEffect)(() => {
    init();
  }, [initialSeconds]);
  (0, _react.useEffect)(() => {
    return () => {
      pause();
    };
  }, []);
  (0, _react.useEffect)(() => {
    if (autoStart) {
      start();
    }
  }, [autoStart]);
  const timer = (0, _react.useCallback)(() => {
    interval.current = _reactNativeBackgroundTimer.default.setInterval(() => {
      if (currentSeconds.current > 0) {
        currentSeconds.current = currentSeconds.current - 1;
        hours.current = ~~(currentSeconds.current / 3600);
        minute.current = ~~(currentSeconds.current % 3600 / 60);
        seconds.current = ~~currentSeconds.current % 60;
        if (onTimes) {
          onTimes(currentSeconds.current);
        }
      }
      if (currentSeconds.current <= 0) {
        if (onEnd) {
          onEnd(currentSeconds.current);
        }
        clear();
      }
      setKey(Math.random());
    }, 1000);
  }, []);
  const start = (0, _react.useCallback)(() => {
    init();
    if (!interval.current) {
      timer();
    }
  }, []);
  const clear = () => {
    if (interval.current) {
      _reactNativeBackgroundTimer.default.clearInterval(interval.current);
      interval.current = 0;
    }
  };
  const font = () => {
    if (fontFamily) {
      return {
        fontFamily: fontFamily
      };
    } else {
      return {};
    }
  };
  const renderTimer = () => {
    if (formatTime === 'hh:mm:ss') {
      if (hours.current > 0) {
        return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
          style: [_styles.styles.text, textStyle, font()]
        }, `${hours.current}:${minute.current.toString().length === 1 ? '0' : ''}${minute.current}:${seconds.current.toString().length === 1 ? '0' : ''}${seconds.current}`);
      } else {
        if (minute.current > 0) {
          return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
            style: [_styles.styles.text, textStyle, font()]
          }, `${minute.current}:${seconds.current.toString().length === 1 ? '0' : ''}${seconds.current}`);
        } else {
          return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
            style: [_styles.styles.text, textStyle, font()]
          }, `${seconds.current}`);
        }
      }
    } else {
      return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: [_styles.styles.text, textStyle, font()]
      }, `${currentSeconds.current}`);
    }
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: style,
    key: key
  }, renderTimer());
});
CountdownComponent.defaultProps = defaultProps;
var _default = CountdownComponent;
exports.default = _default;
